File: android/app/src/main/java/io/github/cyancity/easyunlocker/data/BrokerClient.kt
package io.github.cyancity.easyunlocker.data

import org.json.JSONArray
import org.json.JSONObject
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL

class BrokerClient(
    var brokerUrl: String,
    var deviceToken: String,
) {
    fun pair(pairingToken: String, name: String, vaultId: String = ""): String {
        val body = JSONObject().put("name", name).put("vault_id", vaultId).toString()
        val raw = request("POST", "/v1/device/pair", body, pairingToken)
        return JSONObject(raw).getString("device_token")
    }

    fun pending(): List<PendingRequest> {
        val raw = request("GET", "/v1/device/pending", null, deviceToken)
        val arr = JSONArray(raw)
        return buildList {
            for (i in 0 until arr.length()) {
                val o = arr.getJSONObject(i)
                add(
                    PendingRequest(
                        request_id = o.optString("request_id"),
                        item = o.optString("item"),
                        mode = o.optString("mode"),
                        purpose = o.optString("purpose"),
                        ttl = o.optInt("ttl"),
                        target = o.optString("target"),
                        requester = o.optString("requester"),
                        expires_at = o.optString("expires_at"),
                        state = o.optString("state"),
                        seal_public_key = o.optString("seal_public_key"),
                        delivery = o.optString("delivery"),
                        public_key = o.optString("public_key"),
                        ssh_user = o.optString("ssh_user"),
                        cert_ttl = o.optInt("cert_ttl"),
                    )
                )
            }
        }
    }

    fun registerPushToken(pushToken: String) {
        if (pushToken.isBlank()) return
        request("POST", "/v1/device/push-token", JSONObject().put("token", pushToken).toString(), deviceToken)
    }

    fun decide(requestId: String, decision: String, payload: String?) {
        val body = JSONObject().put("decision", decision)
        if (!payload.isNullOrBlank()) body.put("payload", payload)
        request("POST", "/v1/decision/$requestId", body.toString(), deviceToken)
    }

    /** 生成一次性配对码（10 分钟），给新机器的 `easyGet pair` 用。返回 码 to 过期时刻（epoch 毫秒）。 */
    fun createPairCode(): Pair<String, Long> {
        val raw = request("POST", "/v1/device/pair-code", "{}", deviceToken)
        val o = JSONObject(raw)
        val expiresAt = runCatching { java.time.Instant.parse(o.getString("expires_at")).toEpochMilli() }.getOrDefault(0L)
        return o.getString("code") to expiresAt
    }

    /** 已配对设备列表（不含令牌本体）。 */
    fun devices(): List<PairedDevice> {
        val raw = request("GET", "/v1/device/devices", null, deviceToken)
        val arr = JSONObject(raw).optJSONArray("devices") ?: JSONArray()
        return buildList {
            for (i in 0 until arr.length()) {
                val o = arr.getJSONObject(i)
                add(
                    PairedDevice(
                        id = o.optString("id"),
                        name = o.optString("name"),
                        createdAt = o.optString("created_at"),
                        lastUsedAt = if (o.isNull("last_used_at")) "" else o.optString("last_used_at"),
                        expiresAt = o.optString("expires_at"),
                        current = o.optBoolean("current"),
                    )
                )
            }
        }
    }

    /** 撤销一台设备（可以是当前这台 = 登出）。 */
    fun revokeDevice(id: String) {
        request("POST", "/v1/device/revoke", JSONObject().put("id", id).toString(), deviceToken)
    }

    /** 给一台设备改名（同租户内，可以是当前这台）。 */
    fun renameDevice(id: String, name: String) {
        request("POST", "/v1/device/rename", JSONObject().put("id", id).put("name", name).toString(), deviceToken)
    }

    /** 给当前设备续期 180 天，返回新的过期时间。 */
    fun renewDevice(): String {
        val raw = request("POST", "/v1/device/renew", "{}", deviceToken)
        return JSONObject(raw).optString("expires_at")
    }

    private fun request(method: String, path: String, body: String?, bearer: String): String {
        val url = URL(brokerUrl.trimEnd('/') + path)
        val conn = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = method
            connectTimeout = 15000
            readTimeout = 15000
            setRequestProperty("Authorization", "Bearer $bearer")
            if (body != null) {
                doOutput = true
                setRequestProperty("Content-Type", "application/json")
            }
        }
        if (body != null) {
            OutputStreamWriter(conn.outputStream, Charsets.UTF_8).use { it.write(body) }
        }
        val code = conn.responseCode
        val stream = if (code in 200..299) conn.inputStream else conn.errorStream
        val text = stream.bufferedReader().readText()
        if (code !in 200..299) throw RuntimeException("Broker $code $text")
        return text
    }
}

File: ios/EasyUnlocker/Data/BrokerClient.swift
import Foundation

/// 与安卓 BrokerClient 同一组 REST 端点，报文字段一字不改。
final class BrokerClient {
    let brokerUrl: String
    let deviceToken: String

    init(brokerUrl: String, deviceToken: String) {
        self.brokerUrl = brokerUrl
        self.deviceToken = deviceToken
    }

    struct BrokerError: LocalizedError {
        let code: Int
        let body: String
        var errorDescription: String? { "Broker \(code) \(body)" }
    }

    func pair(pairingToken: String, name: String, vaultId: String = "") async throws -> String {
        let raw = try await request("POST", "/v1/device/pair", ["name": name, "vault_id": vaultId], bearer: pairingToken)
        return raw["device_token"] as? String ?? ""
    }

    func pending() async throws -> [PendingRequest] {
        let data = try await requestRaw("GET", "/v1/device/pending", nil, bearer: deviceToken)
        let arr = (try? JSONSerialization.jsonObject(with: data)) as? [[String: Any]] ?? []
        return arr.map { o in
            var r = PendingRequest()
            r.requestId = o["request_id"] as? String ?? ""
            r.item = o["item"] as? String ?? ""
            r.mode = o["mode"] as? String ?? ""
            r.purpose = o["purpose"] as? String ?? ""
            r.ttl = (o["ttl"] as? NSNumber)?.intValue ?? 0
            r.target = o["target"] as? String ?? ""
            r.requester = o["requester"] as? String ?? ""
            r.expiresAt = o["expires_at"] as? String ?? ""
            r.state = o["state"] as? String ?? ""
            r.sealPublicKey = o["seal_public_key"] as? String ?? ""
            r.delivery = o["delivery"] as? String ?? ""
            r.publicKey = o["public_key"] as? String ?? ""
            r.sshUser = o["ssh_user"] as? String ?? ""
            r.certTtl = (o["cert_ttl"] as? NSNumber)?.intValue ?? 0
            return r
        }
    }

    func registerPushToken(_ pushToken: String) async throws {
        if pushToken.isEmpty { return }
        _ = try await request("POST", "/v1/device/push-token", ["token": pushToken], bearer: deviceToken)
    }

    func decide(requestId: String, decision: String, payload: String?) async throws {
        var body: [String: Any] = ["decision": decision]
        if let payload, !payload.isEmpty { body["payload"] = payload }
        _ = try await request("POST", "/v1/decision/\(requestId)", body, bearer: deviceToken)
    }

    /// 生成一次性配对码（10 分钟），给新机器的 `easyGet pair` 用。
    func createPairCode() async throws -> (code: String, expiresAt: Int64) {
        let raw = try await request("POST", "/v1/device/pair-code", [:], bearer: deviceToken)
        let code = raw["code"] as? String ?? ""
        var expiresAt: Int64 = 0
        if let s = raw["expires_at"] as? String,
           let d = ISO8601DateFormatter().date(from: s) {
            expiresAt = Int64(d.timeIntervalSince1970 * 1000)
        }
        return (code, expiresAt)
    }

    /// 已配对设备列表（不含令牌本体）。
    func devices() async throws -> [PairedDevice] {
        let raw = try await request("GET", "/v1/device/devices", nil, bearer: deviceToken)
        let arr = raw["devices"] as? [[String: Any]] ?? []
        return arr.map { o in
            PairedDevice(
                id: o["id"] as? String ?? "",
                name: o["name"] as? String ?? "",
                createdAt: o["created_at"] as? String ?? "",
                lastUsedAt: o["last_used_at"] as? String ?? "",
                expiresAt: o["expires_at"] as? String ?? "",
                current: o["current"] as? Bool ?? false
            )
        }
    }

    /// 撤销一台设备（可以是当前这台 = 登出）。
    func revokeDevice(id: String) async throws {
        _ = try await request("POST", "/v1/device/revoke", ["id": id], bearer: deviceToken)
    }

    /// 给当前设备续期 180 天，返回新的过期时间。
    func renewDevice() async throws -> String {
        let raw = try await request("POST", "/v1/device/renew", [:], bearer: deviceToken)
        return raw["expires_at"] as? String ?? ""
    }

    /// 给一台设备改名（同租户内，可以是当前这台）。
    func renameDevice(id: String, name: String) async throws {
        _ = try await request("POST", "/v1/device/rename", ["id": id, "name": name], bearer: deviceToken)
    }

    // MARK: - 底层

    @discardableResult
    private func request(_ method: String, _ path: String, _ body: [String: Any]?, bearer: String) async throws -> [String: Any] {
        let data = try await requestRaw(method, path, body, bearer: bearer)
        return (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] ?? [:]
    }

    private func requestRaw(_ method: String, _ path: String, _ body: [String: Any]?, bearer: String) async throws -> Data {
        let url = URL(string: brokerUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/")) + path)!
        var req = URLRequest(url: url, timeoutInterval: 15)
        req.httpMethod = method
        req.setValue("Bearer \(bearer)", forHTTPHeaderField: "Authorization")
        if let body {
            req.httpBody = try JSONSerialization.data(withJSONObject: body)
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }
        let (data, resp) = try await URLSession.shared.data(for: req)
        let code = (resp as? HTTPURLResponse)?.statusCode ?? -1
        guard (200...299).contains(code) else {
            throw BrokerError(code: code, body: String(decoding: data, as: UTF8.self))
        }
        return data
    }
}
